lib/location/method_gnss.c

```diff
@@ -0,0 +1,835 @@
+		method_gnss_cancel();
+		loc_core_event_cb(&location_result);
+	}
+}
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+static void method_gnss_agps_cb_work_fn(struct k_work *item)
+{
+	loc_core_event_cb_assistance_request(&agps_request);
+}
+#endif
+
+static void method_gnss_positioning_work_fn(struct k_work *work)
+{
+	int err = 0;
+	struct k_work_args *work_data = CONTAINER_OF(work, struct k_work_args, work_item);
+	const struct loc_gnss_config gnss_config = work_data->gnss_config;
+
+	if (!method_gnss_allowed_to_start()) {
+		/* Location request was cancelled while waiting for RRC idle or PSM. Do nothing. */
+		return;
+	}
+
+	/* Configure GNSS to continuous tracking mode */
+	err = nrf_modem_gnss_fix_interval_set(1);
+
+	/* By default we take the first fix. */
+	fixes_remaining = 1;
+
+	uint8_t use_case;
+
+	switch (gnss_config.accuracy) {
+	case LOC_ACCURACY_LOW:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START |
+			   NRF_MODEM_GNSS_USE_CASE_LOW_ACCURACY;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOC_ACCURACY_NORMAL:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOC_ACCURACY_HIGH:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+
+		/* In high accuracy mode, use the configured fix count. */
+		fixes_remaining = gnss_config.num_consecutive_fixes;
+		break;
+	}
+
+	if (err) {
+		LOG_ERR("Failed to configure GNSS");
+		loc_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	err = nrf_modem_gnss_start();
+	if (err) {
+		LOG_ERR("Failed to start GNSS");
+		loc_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	loc_core_timer_start(gnss_config.timeout);
+}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void method_gnss_date_time_event_handler(const struct date_time_evt *evt)
+{
+	switch (evt->type) {
+	case DATE_TIME_OBTAINED_MODEM:
+		LOG_DBG("DATE_TIME_OBTAINED_MODEM");
+		break;
+	case DATE_TIME_OBTAINED_NTP:
+		LOG_DBG("DATE_TIME_OBTAINED_NTP");
+		break;
+	case DATE_TIME_OBTAINED_EXT:
+		LOG_DBG("DATE_TIME_OBTAINED_EXT");
+		break;
+	case DATE_TIME_NOT_OBTAINED:
+		LOG_DBG("DATE_TIME_NOT_OBTAINED");
+		break;
+	default:
+		break;
+	}
+}
+#endif
+
+int method_gnss_location_get(const struct loc_method_config *config)
+{
+	const struct loc_gnss_config gnss_config = config->gnss;
+	int err;
+
+	if (running) {
+		LOG_ERR("Previous operation ongoing.");
+		return -EBUSY;
+	}
+
+	/* GNSS event handler is already set once in method_gnss_init(). If no other thread is
+	 * using GNSS, setting it again is not needed.
+	 */
+	err = nrf_modem_gnss_event_handler_set(method_gnss_event_handler);
+	if (err) {
+		LOG_ERR("Failed to set GNSS event handler, error %d", err);
+		return -err;
+	}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	if (!date_time_is_valid()) {
+		date_time_update_async(method_gnss_date_time_event_handler);
```

Comment:
IMHO, this should be handled by the sample/application. If the Location library needs time to be set and it isn't, it should return an error. The time should be updated from a single place and I think that needs to be the application.

Reply:
I agree with @tokangas on this matter. And nrf_cloud_pgps does print and error if time is not set. I think it'll continue nevertheless, but at least there is some indication.

Reply:
So, what do we do with this? BTW, does the code above actually work, because it doesn't wait for the time to be synchronized before initializing P-GPS? I think you should wait for an event from the date_time lib. But anyway, that doesn't matter if we decide to remove this.

Reply:
We'll remove `date_time_update_async` from here as discussed. This should be handled by date_time lib when we have time to implement it and for the time being it's handled by location sample / modem_shell.

Reply:
I think we should do something about PGPS if it continues some process even if time is invalid.

Reply:
Agree. I'll remove it now  from here and put improving PGPS onto my todo list.